File: src/web/invaders/scripts/entities/player.js
// This library is distributed in  the hope that it will be useful but without
// any  warranty; without  even  the  implied  warranty of  merchantability or
// fitness for a particular purpose.
// The use and distribution terms for this software are covered by the Eclipse
// Public License 1.0  (http://opensource.org/licenses/eclipse-1.0.php)  which
// can be found in the file epl-v10.html at the root of this distribution.
// By using this software in any  fashion, you are agreeing to be bound by the
// terms of this license. You  must not remove this notice, or any other, from
// this software.
// Copyright (c) 2013 Cherimoia, LLC. All rights reserved.

(function(undef) { "use strict"; var global = this, _ = global._ ;

var asterix = global.ZotohLab.Asterix,
sh = global.ZotohLab.Asterix,
ccsx = asterix.COCOS2DX,
ivs = asterix.Invaders,
sjs= global.SkaroJS;


//////////////////////////////////////////////////////////////////////////////
// module def
//////////////////////////////////////////////////////////////////////////////

var Ship = cc.Sprite.extend({

  coolDown: function() {
    this.setSpriteFrame(this.frame0);
    this.ammoCount= 0;
  },

  loadAmmo: function() {
    this.setSpriteFrame(this.frame1);
    this.ammoCount= 1;
  },

  hasAmmo: function() {
    return this.ammoCount > 0;
  },

  ammoCount: 0,

  ctor: function(x,y,options) {
    this.options= options;
    this.frame0 = ccsx.getSpriteFrame(options.frames[0]);
    this.frame1 = ccsx.getSpriteFrame(options.frames[1]);
    this._super();
    this.loadAmmo();
    this.setPosition(x,y);
  }

});


asterix.Invaders.EntityPlayer = asterix.XEntity.xtends({

  create: function() {
    return this.sprite = new Ship(this.startPos.x, this.startPos.y, this.options);
  },

  update: function(dt) {
    if (this.sprite) {
      if (ccsx.timerDone(this.coolAmmo)) {
        this.sprite.loadAmmo();
      }
      this._super(dt);
    }
  },

  keypressed: function(dt) {
    if (sh.main.keyboard[cc.KEY.right]) {
      this.onRight(dt);
    }
    if (sh.main.keyboard[cc.KEY.left]) {
      this.onLeft(dt);
    }
    if (sh.main.keyboard[cc.KEY.space] && this.sprite.hasAmmo()) {
      this.doFire();
    }
  },

  doFire: function() {
    var pos = this.sprite.getPosition();
    sh.fireEvent('/game/objects/players/shoot', {
      x: pos.x , y: pos.y + 4
    });
    this.sprite.coolDown();
    this.coolAmmo = ccsx.createTimer(this.sprite, this.options.coolDown);
  },

  onRight: function(dt) {
    this.vel.x = 150;
    this.move(dt);
    this.clamp();
  },

  onLeft: function(dt) {
    this.vel.x = -150;
    this.move(dt);
    this.clamp();
  },

  clamp: function() {
    var sz= this.sprite.getContentSize(),
    pos= this.sprite.getPosition(),
    csts = sh.xcfg.csts,
    wz = ccsx.screen();

    if (ccsx.getRight(this.sprite) > wz.width - csts.TILE) {
      this.sprite.setPosition(wz.width - csts.TILE - sz.width * 0.5, pos.y);
    }
    if (ccsx.getLeft(this.sprite) < csts.TILE) {
      this.sprite.setPosition( csts.TILE + sz.width * 0.5, pos.y);
    }
  },

  check: function(other) {
    other.injured(0,this);
    this.injured(0,other);
  },

  injured: function(num, from) {
    this.dispose();
    sh.fireEvent('/game/objects/players/killed');
  },

  ctor: function(x, y, options) {
    this._super(x, y, options);
    this.vel.x = 150;
    this.options.frames= [ 'ship_0.png', 'ship_1.png' ];
  }


});




}).call(this);

//////////////////////////////////////////////////////////////////////////////
//EOF

File: src/main/clojure/czlab/odin/game/player.clj
;; This library is distributed in  the hope that it will be useful but without
;; any  warranty; without  even  the  implied  warranty of  merchantability or
;; fitness for a particular purpose.
;; The use and distribution terms for this software are covered by the Eclipse
;; Public License 1.0  (http://opensource.org/licenses/eclipse-1.0.php)  which
;; can be found in the file epl-v10.html at the root of this distribution.
;; By using this software in any  fashion, you are agreeing to be bound by the
;; terms of this license. You  must not remove this notice, or any other, from
;; this software.
;; Copyright (c) 2013-2015, Ken Leung. All rights reserved.

(ns ^{:doc ""
      :author "kenl" }

  czlab.odin.game.player

  (:require
    [czlab.xlib.util.core :refer [MubleObj! ]]
    [czlab.xlib.util.logging :as log]
    [czlab.xlib.util.str :refer [strim hgl?]])

  (:import
    [com.zotohlab.odin.game Game PlayRoom
    Player PlayerSession]
    [com.zotohlab.skaro.core Muble]
    [com.zotohlab.odin.core Session]))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;(set! *warn-on-reflection* true)


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; players
;; { player-id -> player }
(def ^:private PLAYER-REGO (ref {}))

;; { player-id -> map of sessions { id -> session } }
(def ^:private PLAYER-SESS (ref {}))
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn ReifyPlayer ""

  ^Player
  [^String user ^String pwd]

  (let [impl (MubleObj!) ]
    (reify

      Player

      (setEmailId [_ email] (.setv impl :email email))
      (emailId [_] (.getv impl :email))

      (setName [_ n] (.setv impl :name n))
      (getName [_] (.getv impl :name))
      (id [_] user)

      (removeSession [_ ps]
        (dosync
          (when-some [m (@PLAYER-SESS user) ]
            (alter PLAYER-SESS
                   assoc
                   user
                   (dissoc m (.id ^Session ps))))))

      (countSessions [_]
        (dosync
          (if-some [m (@PLAYER-SESS user)]
            (int (count m))
            (int 0))))

      (addSession [_ ps]
        (dosync
          (let [m (or (@PLAYER-SESS user) {}) ]
            (alter PLAYER-SESS
                   assoc
                   user
                   (assoc m (.id ^Session ps) ps)))))

      (logout [_ ps]
        (dosync
          (when-some [m (@PLAYER-SESS user)]
            (doseq [^PlayerSession
                    pss (vals m)]
              (.close pss))
            (alter PLAYER-SESS dissoc user)))))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn RemovePlayer ""

  ^Player
  [^String user]

  (dosync
    (when-some [p (@PLAYER-REGO user) ]
      (alter PLAYER-REGO dissoc user)
      (when-some [m (@PLAYER-SESS user)]
        (doseq [^Session v (vals m)]
          (.close v)))
      (alter PLAYER-SESS dissoc user)
      p)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn CreatePlayer ""

  ^Player
  [^String user ^String pwd]

  (dosync
    (if-some [p (@PLAYER-REGO user)]
      p
      (let [p2 (ReifyPlayer user pwd)]
        (alter PLAYER-REGO assoc user p2)
        (alter PLAYER-SESS assoc user {})
        p2))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
(defn LookupPlayer ""

  (^Player
    [^String user ^String pwd]
    (CreatePlayer user pwd))

  (^Player
    [^String user]
    (dosync (get @PLAYER-REGO user))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;EOF

